pyccel/ast/headers.py

```diff
@@ -269,14 +277,14 @@ def process_template(arg_code, Tname, d_type, type_index):
 
         #TODO: handle the case of functions arguments
 
-        arg_codes = [list(arg_code) for arg_code in product(*dtypes)]
-        for tmplt in templates:
-            arg_codes = [process_template(arg_code, tmplt, d_type, j) for arg_code in arg_codes\
-                    for j, d_type, in enumerate(templates[tmplt].args) if arg_code]
-        if len(arg_codes) > 1:
-            arg_codes = list(filter(None, arg_codes))
+        signatures = list(product(*dtypes))
+        new_templates = find_templates(signatures[0], templates)
+
+        for tmplt in new_templates:
+            signatures = [process_template(s, tmplt, d_type)\
+                    for s in signatures for d_type in new_templates[tmplt].args]
```

Comment:
This looks very concise and elegant now!

I suggest you make `signatures` a tuple instead of a list, because a `tuple` is an immutable object (it has no `__setitem__` method) and hence we make it clear that we never change it in our algorithm. What we do is to define a **new** tuple at each loop cycle instead.